File: mergesort/mergesort.py
import time
import random

def msort(A):
    mergesort(A, 0, len(A)) 
def mergesort(A, p, r):
    if p < r - 1:
        q = (p + r) // 2
        mergesort(A, p, q)
        mergesort(A, q, r)
        merge(A, p, q, r)

def merge(A, p, q, r):
    left = A[p:q]
    right = A[q:r]
    
    sentinel = 1001
    left.append(sentinel)
    right.append(sentinel)

    i, j = 0, 0
    for k in range(p, r):
        if left[i] <= right[j]:
            A[k] = left[i]
            i += 1
        else:
            A[k] = right[j]
            j += 1

a = [1,2,3,4,5,0,9,7,3]
msort(a)
print(a)

n = 10000
b = []
for i in range(n):
    r = random.randint(0, 1000)
    b.append(r)
print("Begin sorting")
begin = time.time_ns()
msort(b)
end = time.time_ns()
print("merge sort finished sorting {} elements in {}ns".format(n, (end - begin)))

File: MergeSort/Mergesort.jl
module Mergesort 
    export mergesort!

    NumOrChar = Union{Number, Char}

    function merge!(arr::Array{T, 1}, p::Int, m::Int, q::Int) where (T <: NumOrChar)
        n₁ = m - p + 1
        n₂ = q - m

        # Create temporary arrays 
        left = Array{T, 1}(undef, n₁ + 1)
        right = Array{T, 1}(undef, n₂ + 1)

        # Save everything from the left side of the array to the left array 
        # left[1:n₁] .= arr[p : p + n₁ - 1]
        # right[1:n₂] .= arr[m : m + n₂ - 1]

        for i ∈ 1:n₁
            left[i] = arr[p + i - 1] 
        end

        for j ∈ 1:n₂
            right[j] = arr[m + j]
        end

        # Create sentinel values
        # The for loop later will never reach this because
        # we only have to look at p-q elements while the arrays has (p-q)+2 elements.
        left[n₁ + 1] = typemax(T)
        right[n₂ + 1] = typemax(T)

        i, j = (1, 1)
        
        # for each k in {p, p + 1,…, q}
        for k ∈ p:q
            if left[i] ≤ right[j]
                arr[k] = left[i]
                i += 1
            else
                arr[k] = right[j]
                j += 1
            end
        end
    end

    function msort!(arr::Array{T, 1}, p::Int, q::Int) where (T <: NumOrChar)
        if p < q 
            mid = (p + q) ÷ 2

            msort!(arr, p, mid)
            msort!(arr, mid+1, q)
            merge!(arr, p, mid, q)
        end
    end

    function mergesort!(arr::Array{T, 1}) where (T <: NumOrChar)
        msort!(arr, 1, length(arr))
    end
end